reverseproxy: Implement modular CA provider for TLS transport

armadi1809 opened the pull request:
Closes #6063

armadi1809 commented:
Looks like the failed test is just a server timeout.

mohammed90 commented on modules/caddyhttp/reverseproxy/httptransport.go:
```diff
@@ -564,6 +568,18 @@ func (t TLSConfig) MakeTLSClientConfig(ctx caddy.Context) (*tls.Config, error) {
+		if !ok {
+			return nil, fmt.Errorf("CARaw module '%s' is not a certificate pool provider", ca)
+		}
+		cfg.ClientCAs = ca.CertPool()
```
The `ClientCAs` field is for servers who are validating incoming client connections. In reverse-proxy transport, we're doing the opposite. We need the block above to be edited, i.e. modifying the source of root certificates.

armadi1809 replied:
Sorry, forgot to change it from the code I ported from your PR. I fixed this. I also added a check to make sure we are not using the old and the new methods of getting the `RootCAs` at the same time. Thanks for the feedback.

mohammed90 commented on modules/caddyhttp/reverseproxy/httptransport.go:
```diff
@@ -564,6 +568,21 @@ func (t TLSConfig) MakeTLSClientConfig(ctx caddy.Context) (*tls.Config, error) {
+		}
+		ca, ok := caRaw.(caddytls.CA)
+		if !ok {
+			return nil, fmt.Errorf("CARaw module '%s' is not a certificate pool provider", ca)
```
Use `ca` instead of `CARaw` because the user does not know the field name inside the struct.

armadi1809 replied:
I am assuming you are talking about the error message here right? If so, then this should now be resolved. Also, I think you are using the same error message when provisioning the `ClientAuthentication` module inside `modules/caddytls/connpolicy.go`, so maybe we should also change that? 

Sorry in advance if it's not about the error message and if  I misinterpreted your comment 😅

mohammed90 replied:
> so maybe we should also change that?

Yep. It was probably an auto-complete mistake, and we missed it.

mohammed90 reviewed (changes requested):
GitHub does not let me comment on the unchanged lines, so excuse the non-inlined comments.

Mark the old way as deprecated and subject to removal in the future, and emit logs saying the same when used. You can get a logger from `ctx.Logger()`.

armadi1809 commented:
I just added the "deprecated" comments to `RootCAPool` and `RootCAPEMFiles` fields as well as a deprecation warning log when those are used.

mohammed90 reviewed (changes requested):
Thanks for doing the rounds, Aziz! Sorry for not giving the review points in full at once. I'm squeezing the review time in tight time slots 😅 

We're missing tests and Caddyfile support

mohammed90 commented on modules/caddyhttp/reverseproxy/httptransport.go:
```diff
@@ -546,6 +552,7 @@ func (t TLSConfig) MakeTLSClientConfig(ctx caddy.Context) (*tls.Config, error) {
 
 	// trusted root CAs
 	if len(t.RootCAPool) > 0 || len(t.RootCAPEMFiles) > 0 {
+		ctx.Logger().Warn("RootCAPool and RootCAPEMFiles are deprecated. Use one of the tls.ca_pool.source modules instead")
```
Same for the fields names here. The users only know `root_ca_pool` and `root_ca_pem_files`.

armadi1809 commented:
No worries at all Mohammed! Thank you for your feedback. I implemented the caddy file support and added both caddy file integration tests as well as tests for the `(h *HTTPTransport) UnmarshalCaddyfile` function. Please let me know if anything else is still needed here. Thanks again!

mholt reviewed (approved):
Sorry for falling behind -- this is looking better now! I'd be willing to give this a try. Thank you!!

mholt commented:
@mohammed90 Anything you can think of we need to change for this?

mohammed90 commented on modules/caddyhttp/reverseproxy/caddyfile.go:
```diff
@@ -1218,6 +1222,28 @@ func (h *HTTPTransport) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+			if len(h.TLS.RootCAPEMFiles) != 0 {
+				return d.Err("cannot specify both 'tls_trust_pool' and 'tls_trusted_ca_certs'")
+			}
+			h.TLS.CARaw = caddyconfig.JSONModuleObject(ca, "provider", modStem, nil)
```
I think we should return an error if `h.TLS.CARaw` is not nil because it means they've specified it twice.

mohammed90 commented on modules/caddyhttp/reverseproxy/httptransport.go:
```diff
@@ -564,6 +571,21 @@ func (t TLSConfig) MakeTLSClientConfig(ctx caddy.Context) (*tls.Config, error) {
 
+	if t.CARaw != nil {
+		if len(t.RootCAPool) > 0 || len(t.RootCAPEMFiles) > 0 {
+			return nil, fmt.Errorf("Conflicting config for Root CA pool")
```
```suggestion
			return nil, fmt.Errorf("conflicting config for Root CA pool")
```

mohammed90 commented on modules/caddyhttp/reverseproxy/httptransport_test.go:
```diff
@@ -0,0 +1,97 @@
+			args: args{
+				d: caddyfile.NewTestDispenser(
+					`http {
+					tls_trust_pool inline MIIDSzCCAjOgAwIBAgIUfIRObjWNUA4jxQ/0x8BOCvE2Vw4wDQYJKoZIhvcNAQELBQAwFjEUMBIGA1UEAwwLRWFzeS1SU0EgQ0EwHhcNMTkwODI4MTYyNTU5WhcNMjkwODI1MTYyNTU5WjAWMRQwEgYDVQQDDAtFYXN5LVJTQSBDQTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAK5m5elxhQfMp/3aVJ4JnpN9PUSz6LlP6LePAPFU7gqohVVFVtDkChJAG3FNkNQNlieVTja/bgH9IcC6oKbROwdY1h0MvNV8AHHigvl03WuJD8g2ReVFXXwsnrPmKXCFzQyMI6TYk3m2gYrXsZOU1GLnfMRC ...
+					tls_trusted_ca_certs MIIDSzCCAjOgAwIBAgIUfIRObjWNUA4jxQ/0x8BOCvE2Vw4wDQYJKoZIhvcNAQELBQAwFjEUMBIGA1UEAwwLRWFzeS1SU0EgQ0EwHhcNMTkwODI4MTYyNTU5WhcNMjkwODI1MTYyNTU5WjAWMRQwEgYDVQQDDAtFYXN5LVJTQSBDQTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAK5m5elxhQfMp/3aVJ4JnpN9PUSz6LlP6LePAPFU7gqohVVFVtDkChJAG3FNkNQNlieVTja/bgH9IcC6oKbROwdY1h0MvNV8AHHigvl03WuJD8g2ReVFXXwsnrPmKXCFzQyMI6TYk3m2gYrXsZOU1GLnfMRC3 ...
```
Replace the DER with the variable `test_der_1` like you did below

mohammed90 reviewed (changes requested):
A few minor nitpicks and I believe I got everything

mohammed90 commented on modules/caddyhttp/reverseproxy/caddyfile.go:
```diff
@@ -1218,6 +1222,31 @@ func (h *HTTPTransport) UnmarshalCaddyfile(d *caddyfile.Dispenser) error {
+				return d.Err("cannot specify both 'tls_trust_pool' and 'tls_trusted_ca_certs'")
+			}
+			if h.TLS.CARaw != nil {
+				return d.Err("cannot specify \"ca\" field twice in tls_trust_pool")
```
They're setting `tls_trust_pool`, not `ca`. In Caddyfile format, they don't see the `ca` field of the JSON. They see `tls_trust_pool`.

armadi1809 commented:
@mohammed90 I think all your feedback items have been resolved Thanks for reviewing!

mohammed90 reviewed (approved):
LGTM! Waiting on @mholt review then we're good to go.
